### PR TITLE
feat: add codesign verification and assessment prior to notarizing

### PR DESCRIPTION
**Pull request by jonluca**
Prior to stapling, we should run an `spctl --assess` and `codesign` check to verify that the stapling will work. This caused major issues for us twice - electron builder would modify some files in the build, and that was causing stapling and notarization to fail. This will verify that the app is signed properly.

One potential change is to only run these if stapling fails - open to suggestions on that.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -2,3 +2,4 @@ node_modules
 lib
 *.log
 src/example.ts
+.idea
```
Comment (dsanders11) [inline .gitignore:5]: Can you remove this change? I believe our general stance at the moment is IDE-specific ignores shouldn't go in the committed `.gitignore`.

File: src/index.ts [hunk 2]
```diff
@@ -50,6 +51,8 @@ export async function notarize({ appPath, ...otherOptions }: NotarizeOptions) {
     } as NotaryToolStartOptions);
   }
 
+  await checkSignatures({ appPath });
```
Comment (dsanders11) [inline src/index.ts:54]: Should this go before notarizing, rather than before stapling? Notarization will fail if the signatures are invalid.

**Comment by erickzhao** [comment]
Merged in the main branch to try to kick CI

File: src/index.ts (hunk 2 above)
Comment (jonluca) [reply src/index.ts]: Yes great point

File: .gitignore (hunk 1 above)
Comment (jonluca) [reply .gitignore]: Will do

File: src/check-signature.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
+import * as path from 'path';
+
+import { spawn } from './spawn';
+import type { NotarizeStapleOptions } from './types';
```
Comment (dsanders11) [inline src/check-signature.ts:4]:
```suggestion
import { NotarizeStapleOptions } from './types';
```

**Comment by dsanders11** [comment]
Looks like lint is failing. I also pushed a change to turn an `import type` into just `import` - I think the version of prettier on this repo is too old to understand that syntax.

**Comment by dsanders11** [comment]
@jonluca, I think you need to log out of CircleCI and back in again, then push something to trigger CI (empty commit should do it). [This is the CircleCI run associated with your last push](https://app.circleci.com/pipelines/github/electron/notarize/198). Happens from time to time with CircleCI unfortunately.

> ⚠️ Could not find a usable config.yml, you may have revoked the CircleCI OAuth app.
Please sign out of CircleCI and log back in with your VCS before triggering a new pipeline.

**Comment by jonluca** [comment]
done!

File: src/index.ts [hunk 4]
```diff
@@ -6,6 +6,7 @@ import { startLegacyNotarize, waitForLegacyNotarize } from './legacy';
 import { isNotaryToolAvailable, notarizeAndWaitForNotaryTool } from './notarytool';
 import { stapleApp } from './staple';
 import { NotarizeOptions, NotaryToolStartOptions } from './types';
+import { checkSignatures } from './check-signature';
```
Comment (dsanders11) [inline src/index.ts:9]: These are currently alpha sorted, could we put this at the top of this block to retain that?

File: src/check-signature.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+  const result = await spawn(
+    'codesign',
+    ['-vvv', '--deep', '--strict', path.basename(opts.appPath)],
+    {
+      cwd: path.dirname(opts.appPath),
+    },
+  );
+
+  return result;
+};
```
Comment (dsanders11) [inline src/check-signature.ts:26]:
```suggestion
};

```

File: src/check-signature.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+export async function checkSignatures(opts: NotarizeStapleOptions): Promise<void> {
+  const codesignResult = await codesign(opts);
+  const codesignInfo = await codesignDisplay(opts);
```
Comment (dsanders11) [inline src/check-signature.ts:29]:
```suggestion
  const [codesignResult, codesignInfo] = await Promise.all([codesign(opts), codesignDisplay(opts)]);
```

These should be able to be run in parallel, right?

Comment (jonluca) [reply src/check-signature.ts]: yes done

File: src/check-signature.ts (hunk 5 above)
Comment (jonluca) [reply src/check-signature.ts:26]: yes. we should make this an automatic rule in eslint though

**Review by dsanders11 — APPROVED** [review]

**Review by felixrieseberg — APPROVED** [review]
This looks good! 

@jonluca Thanks for adding this :bow:

**Review by VerteDinde — APPROVED** [review]

**Comment by DanielMcAssey** [comment]
This has broken our .pkg signing, it throws an error, where it was working before (Using a patch to sign pkg's and dmg's)

**Comment by jonluca** [comment]
@DanielMcAssey what is the error? If `spctl` or `codesign` return non-zero exit codes that implies there were issues with the bundle, so I'd be interested if it's a warning or something idiosyncratic

I'll prep a revert in the meantime though

**Comment by DanielMcAssey** [comment]
So its a non-zero error, error code 1 `***.pkg: code object is not signed at all`.

**Comment by erickzhao** [comment]
@DanielMcAssey is the patch that you have the same as https://github.com/electron/notarize/pull/154?

**Comment by DanielMcAssey** [comment]
> @DanielMcAssey is the patch that you have the same as #154?

Pretty much, slight variation, it specifically checks for .dmg and .pkg, instead of doing fs.stat

**Comment by dlmartens** [comment]
I'm getting the same error for my signed installer ".pkg." (`code object is not signed at all`)

It looks like codesign cannot verify the signature of an installer package. Without this check, my installer is able to be notarized without a problem. I can verify its signature with `pkgutil --check-signature installer.pkg`, which returns a non-zero exit code on verification error.

This same installer package can be notarized without issue with `"@electron/notarize": "2.1.0"`.

**Pull request merged**